File: seqosclib/src/main/kotlin/ch/bildspur/seqosc/OSCPlayer.kt
package ch.bildspur.seqosc

import ch.bildspur.seqosc.net.OSCClient
import jdk.jshell.spi.ExecutionControl
import java.net.Inet4Address
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import kotlin.concurrent.thread
import kotlin.math.roundToLong


class OSCPlayer(val host : String, val port : Int, var speed : Float = 1.0f, val buffer : OSCBuffer) {

    val client = OSCClient()
    val address = Inet4Address.getByName(host)

    private val playState = AtomicBoolean(false)
    private val index = AtomicInteger(0)

    var lastTimeStamp = 0L
    var loop = false

    lateinit var playThread : Thread

    val currentIndex : Int
        get() = index.get()

    val playing : Boolean
        get() = playState.get()

    fun play() {
        if(playing) return

        playState.set(true)
        playThread = thread(start = true, isDaemon = true) {

            lastTimeStamp = buffer.samples.first().timestamp

            while(playing && index.get() < buffer.samples.size) {
                val sample = buffer.samples[index.getAndIncrement()]
                val delta = sample.timestamp - lastTimeStamp
                Thread.sleep((delta / speed).roundToLong())
                client.send(address, port, sample.packet)

                lastTimeStamp = sample.timestamp

                if(loop && index.get() >= buffer.samples.size) {
                    lastTimeStamp = buffer.samples.first().timestamp
                    index.set(0)
                }
            }

            playState.set(false)
        }
    }

    fun stop() {
        if(!playing) return
        playState.set(false)
        playThread.join()
    }

    fun reset() {
        index.set(0)
    }
}

File: dotnet/OSCPlayer.cs
using System;
using System.Linq;
using System.Net;
using System.Threading;
using SeqOSC.Net;

namespace SeqOSC
{
    public class OSCPlayer
    {
        public OSCBuffer Buffer { get; set; }
        public float Speed { get; set; } = 1.0f;

        public bool Loop { get; set; } = false;
        public string Host { get; set; } = IPAddress.Loopback.ToString();
        public int Port { get; set; } = 8000;

        public OSCClient Client { get; set; }

        public bool IsPlaying => _playing;

        public int Position => _position;

        private volatile bool _playing = false;
        
        private volatile int _position = 0;

        public OSCPlayer()
        {
            Client = new OSCClient();
        }
        
        public OSCPlayer(OSCBuffer buffer) : this()
        {
            Buffer = buffer;
        }

        public void Play()
        {
            if (Buffer.Samples.Count == 0)
                return;
            
            _playing = true;
            _position = 0;

            var receiver = new IPEndPoint(IPAddress.Parse(Host), Port);
            var lastTimeStamp = Buffer.Samples.First().Timestamp;

            while (_position < Buffer.Samples.Count && _playing)
            {
                var sample = Buffer.Samples[_position++];
                var delta = sample.Timestamp - lastTimeStamp;
                
                Thread.Sleep((int)Math.Round(delta / Speed));
                Client.Send(receiver, sample.Packet);

                lastTimeStamp = sample.Timestamp;

                if (Loop && _position >= Buffer.Samples.Count)
                {
                    lastTimeStamp = Buffer.Samples.First().Timestamp;
                    _position = 0;
                }
            }

            _playing = false;
        }

        public void Stop()
        {
            _playing = false;
        }
    }
}